core/src/avm2/object/dictionary_object.rs

```diff
@@ -45,14 +50,94 @@ impl fmt::Debug for DictionaryObject<'_> {
     }
 }
 
+#[derive(Clone)]
+pub enum DictionaryMap<'gc> {
+    Strong(HashMap<Object<'gc>, Value<'gc>, FnvBuildHasher>),
+    Weak(HashMap<WeakObject<'gc>, Value<'gc>, FnvBuildHasher>),
+}
+
+unsafe impl<'gc> Collect for DictionaryMap<'gc> {
+    fn trace(&self, cc: gc_arena::CollectionContext) {
+        match self {
+            Self::Strong(m) => {
+                for (k, v) in m {
+                    k.trace(cc);
+                    v.trace(cc);
+                }
+            }
+            Self::Weak(m) => {
+                for (k, v) in m {
+                    k.trace(cc);
+                    v.trace(cc);
+                }
+            }
+        }
+    }
+}
+
+impl<'gc> DictionaryMap<'gc> {
+    /// Inserts a value into this Dictionary.
+    pub fn insert(&mut self, obj: Object<'gc>, v: Value<'gc>) {
+        match self {
+            DictionaryMap::Strong(m) => m.insert(obj, v),
+            DictionaryMap::Weak(m) => m.insert(obj.downgrade(), v),
+        };
+    }
+
+    /// Removes a value from this Dictionary.
+    pub fn remove(&mut self, obj: Object<'gc>) {
+        match self {
+            DictionaryMap::Strong(m) => m.remove(&obj),
+            DictionaryMap::Weak(m) => m.remove(&obj.downgrade()),
+        };
+    }
+
+    /// Gets a value in this Dictionary by key.
+    pub fn get(&self, obj: Object<'gc>) -> Option<Value<'gc>> {
+        match self {
+            DictionaryMap::Strong(m) => m.get(&obj).cloned(),
+            DictionaryMap::Weak(m) => m.get(&obj.downgrade()).cloned(),
+        }
+    }
+
+    /// Gets a key at a specific index in this Dictionary
+    ///
+    /// Automatically clears dead references
+    pub fn get_key_at(
+        &mut self,
+        mc: MutationContext<'gc, '_>,
+        index: usize,
+    ) -> Option<Object<'gc>> {
+        self.prune(mc);
+        match self {
+            DictionaryMap::Strong(m) => m.keys().nth(index).cloned(),
+            DictionaryMap::Weak(m) => m.keys().nth(index).and_then(|o| o.upgrade(mc)),
+        }
+    }
+
+    /// Clears out any dead references in this Dictionary
+    pub fn prune(&mut self, mc: MutationContext<'gc, '_>) {
```

Comment:
Should we call this in any other places, to prevent dead keys from staying in the map until you try to get a specific index?